libraries/llvm/buffer.c

```diff
@@ -0,0 +1,141 @@
+
+void c_buffer_destruct(const struct Pos buffer) {
+    free(buffer.obj - BUFFER_METADATA_WIDTH);
+}
+
+
+void c_buffer_refcount_increment(const struct Pos buffer) {
+    if (DEBUG_REFCOUNT) {
+        fprintf(stderr, "c_buffer_refcount_increment((struct Pos) "
+            "{ .tag = %" PRIu64 ", .obj = %p }): from %" PRIu64 "\n",
+            buffer.tag, buffer.obj, *(c_buffer_refcount(buffer)));
+        fflush(stderr);
+    }
+
+    (*c_buffer_refcount(buffer))++;
+}
+
+void c_buffer_refcount_decrement(const struct Pos buffer) {
+    if (DEBUG_REFCOUNT) {
+        fprintf(stderr, "c_buffer_refcount_decrement((struct Pos) "
+            "{ .tag = %" PRIu64 ", .obj = %p }): from %" PRIu64 "\n",
+            buffer.tag, buffer.obj, *(c_buffer_refcount(buffer)));
+        fflush(stderr);
+    }
+
+    if (!(*c_buffer_refcount(buffer))--)
+        c_buffer_destruct(buffer);
+}
+
+
+struct Pos c_buffer_construct_zeroed(const uint64_t n) {
+    uint8_t *zeroes = calloc(n, sizeof *zeroes);
+    ASSERT_NON_NULL(zeroes)
+    const struct Pos buffer = c_buffer_construct(n, zeroes);
+    free(zeroes);
+    return buffer;
+}
+
+void c_buffer_truncate(struct Pos buffer, const uint64_t n) {
+    if (n > c_buffer_length(buffer))
+        return;
+    buffer.tag = n;
+}
+
+struct Pos c_buffer_copy(const struct Pos buffer) {
+    return c_buffer_construct(c_buffer_length(buffer), c_buffer_bytes(buffer));
+}
+
+
+// incurs a malloc: the returned pointer needs to be managed
+char *c_buffer_as_null_terminated_string(const struct Pos buffer) {
+    // Zero runes are represented as non-minimal utf8 in the null-terminated
+    // string. As such, a count reveals the necessary number of bytes.
+    uint64_t zero_runes = 0;
+    for (uint64_t j = 0; j < c_buffer_length(buffer); ++j)
+        zero_runes += !c_buffer_bytes(buffer)[j];
+
+    const uint64_t n = c_buffer_length(buffer) + zero_runes;
+    char *buf = (char *) malloc((n+1) * sizeof *buf);
+    ASSERT_NON_NULL(buf)
+
+    uint64_t i = 0;
+    for (uint64_t j = 0; j < c_buffer_length(buffer); ++j) {
+        buf[i++] = c_buffer_bytes(buffer)[j];
+        if (!buf[i]) {
+            buf[i]   = 0xc0; // 0b110.00000
+            buf[i++] = 0x80; // 0b10.000000
+        }
+    }
+
+    // null-terminated
+    buf[n] = '\00';
+
+    return buf;
+}
+
+struct Pos c_buffer_construct_from_null_terminated_string(const char *data_nt) {
+    uint64_t n = 0;
+    while (data_nt[n++]);
+
+    return c_buffer_construct(n, (uint8_t *) data_nt);
+}
+
+
+struct Pos c_buffer_concatenate(const struct Pos left, const struct Pos right) {
+    const struct Pos concatenated = c_buffer_construct_zeroed(
+        c_buffer_length(left) + c_buffer_length(right));
+    for (uint64_t j = 0; j < c_buffer_length(concatenated); ++j)
+        c_buffer_bytes(concatenated)[j]
+            = j < c_buffer_length(left)
+            ? c_buffer_bytes(left)[j]
+            : c_buffer_bytes(right)[j - c_buffer_length(left)];
+    return concatenated;
```

Comment:
Here we have to `eraseString(left)` and `eraseString(right)`, hmmm. We are already using `memcpy` in `rts.ll`, so not sure...

Reply:
If we do not optimize for singly-owned objects, then any function's arguments *have* to be preserved, so I see no circumvention to eventually erasing both arguments.

Reply:
Ref. your modification [bf2b807](https://github.com/effekt-lang/effekt/pull/142/commits/bf2b807f8c435a79dd7852501fc78c22a51fe796)